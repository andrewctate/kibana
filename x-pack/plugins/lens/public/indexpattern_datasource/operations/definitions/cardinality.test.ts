/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import {
  buildExpression,
  ExpressionAstExpressionBuilder,
  parseExpression,
} from '@kbn/expressions-plugin/common';
import { operationDefinitionMap } from '.';
import { OriginalColumn } from '../../to_expression';

describe('unique values function', () => {
  describe('optimizeEsAggs', () => {
    const buildMapsFromAggBuilders = (aggs: ExpressionAstExpressionBuilder[]) => {
      const esAggsIdMap: Record<string, OriginalColumn[]> = {};
      const aggsToIdsMap = new Map();
      aggs.forEach((builder, i) => {
        const esAggsId = `col-${i}-${i}`;
        esAggsIdMap[esAggsId] = [{ id: `original-${i}` } as OriginalColumn];
        aggsToIdsMap.set(builder, esAggsId);
      });
      return {
        esAggsIdMap,
        aggsToIdsMap,
      };
    };

    it('consolidates duplicate aggs', () => {
      const aggs = [
        // group 1
        'aggCardinality id="0" enabled=true schema="metric" field="bytes" emptyAsNull=false',
        'aggCardinality id="1" enabled=true schema="metric" field="bytes" emptyAsNull=false',
        // group 2
        'aggFilteredMetric id="2" enabled=true schema="metric" \n  customBucket={aggFilter id="2-filter" enabled=true schema="bucket" filter={kql q="geo.dest: \\"GA\\" "}} \n  customMetric={aggCardinality id="2-metric" enabled=true schema="metric" field="bytes" emptyAsNull=false}',
        'aggFilteredMetric id="3" enabled=true schema="metric" \n  customBucket={aggFilter id="3-filter" enabled=true schema="bucket" filter={kql q="geo.dest: \\"GA\\" "}} \n  customMetric={aggCardinality id="3-metric" enabled=true schema="metric" field="bytes" emptyAsNull=false}',
        // group 3
        'aggFilteredMetric id="4" enabled=true schema="metric" \n  customBucket={aggFilter id="4-filter" enabled=true schema="bucket" filter={kql q="geo.dest: \\"AL\\" "}} \n  customMetric={aggCardinality id="4-metric" enabled=true schema="metric" field="bytes" emptyAsNull=false}',
        'aggFilteredMetric id="5" enabled=true schema="metric" \n  customBucket={aggFilter id="5-filter" enabled=true schema="bucket" filter={kql q="geo.dest: \\"AL\\" "}} \n  customMetric={aggCardinality id="5-metric" enabled=true schema="metric" field="bytes" emptyAsNull=false}',
        // group 4
        'aggFilteredMetric id="6" enabled=true schema="metric" \n  customBucket={aggFilter id="6-filter" enabled=true schema="bucket" filter={lucene q="\\"geo.dest: \\\\\\"AL\\\\\\" \\""}} \n  customMetric={aggCardinality id="6-metric" enabled=true schema="metric" field="bytes" emptyAsNull=false}',
        'aggFilteredMetric id="7" enabled=true schema="metric" \n  customBucket={aggFilter id="7-filter" enabled=true schema="bucket" filter={lucene q="\\"geo.dest: \\\\\\"AL\\\\\\" \\""}} \n  customMetric={aggCardinality id="7-metric" enabled=true schema="metric" field="bytes" emptyAsNull=false}',
        // group 5
        'aggFilteredMetric id="8" enabled=true schema="metric" \n  customBucket={aggFilter id="8-filter" enabled=true schema="bucket" filter={kql q="geo.dest: \\"AL\\" "} timeWindow="1m"} \n  customMetric={aggCardinality id="8-metric" enabled=true schema="metric" field="bytes" emptyAsNull=false}',
        'aggFilteredMetric id="9" enabled=true schema="metric" \n  customBucket={aggFilter id="9-filter" enabled=true schema="bucket" filter={kql q="geo.dest: \\"AL\\" "} timeWindow="1m"} \n  customMetric={aggCardinality id="9-metric" enabled=true schema="metric" field="bytes" emptyAsNull=false}',
        // group 6 (unrelated aggs)
        'aggFilteredMetric id="10" enabled=true schema="metric" \n  customBucket={aggFilter id="10-filter" enabled=true schema="bucket" filter={kql q="bytes: *"}} \n  customMetric={aggTopMetrics id="10-metric" enabled=true schema="metric" field="bytes" size=1 sortOrder="desc" sortField="timestamp"}',
        'aggFilteredMetric id="11" enabled=true schema="metric" \n  customBucket={aggFilter id="11-filter" enabled=true schema="bucket" filter={kql q="bytes: *"}} \n  customMetric={aggTopMetrics id="11-metric" enabled=true schema="metric" field="bytes" size=1 sortOrder="desc" sortField="timestamp"}',
      ].map((expression) => buildExpression(parseExpression(expression)));

      const { esAggsIdMap, aggsToIdsMap } = buildMapsFromAggBuilders(aggs);

      const { esAggsIdMap: newIdMap, aggs: newAggs } = operationDefinitionMap.unique_count
        .optimizeEsAggs!(aggs, esAggsIdMap, aggsToIdsMap);

      expect(newAggs).toHaveLength(7);

      expect(newIdMap).toMatchInlineSnapshot(`
        Object {
          "col-0-0": Array [
            Object {
              "id": "original-0",
            },
            Object {
              "id": "original-1",
            },
          ],
          "col-10-10": Array [
            Object {
              "id": "original-10",
            },
          ],
          "col-11-11": Array [
            Object {
              "id": "original-11",
            },
          ],
          "col-2-2": Array [
            Object {
              "id": "original-2",
            },
            Object {
              "id": "original-3",
            },
          ],
          "col-4-4": Array [
            Object {
              "id": "original-4",
            },
            Object {
              "id": "original-5",
            },
          ],
          "col-6-6": Array [
            Object {
              "id": "original-6",
            },
            Object {
              "id": "original-7",
            },
          ],
          "col-8-8": Array [
            Object {
              "id": "original-8",
            },
            Object {
              "id": "original-9",
            },
          ],
        }
      `);

      expect(newAggs.map((agg) => agg.toAst())).toMatchSnapshot();
    });
  });
});
