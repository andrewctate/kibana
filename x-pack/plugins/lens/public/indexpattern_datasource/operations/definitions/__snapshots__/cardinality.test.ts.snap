// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unique values function optimizeEsAggs consolidates duplicate aggs 2`] = `
Array [
  Object {
    "chain": Array [
      Object {
        "arguments": Object {
          "emptyAsNull": Array [
            false,
          ],
          "enabled": Array [
            true,
          ],
          "field": Array [
            "bytes",
          ],
          "id": Array [
            "0",
          ],
          "schema": Array [
            "metric",
          ],
        },
        "function": "aggCardinality",
        "type": "function",
      },
    ],
    "type": "expression",
  },
  Object {
    "chain": Array [
      Object {
        "arguments": Object {
          "customBucket": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "enabled": Array [
                      true,
                    ],
                    "filter": Array [
                      Object {
                        "chain": Array [
                          Object {
                            "arguments": Object {
                              "q": Array [
                                "geo.dest: \\"GA\\" ",
                              ],
                            },
                            "function": "kql",
                            "type": "function",
                          },
                        ],
                        "type": "expression",
                      },
                    ],
                    "id": Array [
                      "2-filter",
                    ],
                    "schema": Array [
                      "bucket",
                    ],
                  },
                  "function": "aggFilter",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "customMetric": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "emptyAsNull": Array [
                      false,
                    ],
                    "enabled": Array [
                      true,
                    ],
                    "field": Array [
                      "bytes",
                    ],
                    "id": Array [
                      "2-metric",
                    ],
                    "schema": Array [
                      "metric",
                    ],
                  },
                  "function": "aggCardinality",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "enabled": Array [
            true,
          ],
          "id": Array [
            "2",
          ],
          "schema": Array [
            "metric",
          ],
        },
        "function": "aggFilteredMetric",
        "type": "function",
      },
    ],
    "type": "expression",
  },
  Object {
    "chain": Array [
      Object {
        "arguments": Object {
          "customBucket": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "enabled": Array [
                      true,
                    ],
                    "filter": Array [
                      Object {
                        "chain": Array [
                          Object {
                            "arguments": Object {
                              "q": Array [
                                "geo.dest: \\"AL\\" ",
                              ],
                            },
                            "function": "kql",
                            "type": "function",
                          },
                        ],
                        "type": "expression",
                      },
                    ],
                    "id": Array [
                      "4-filter",
                    ],
                    "schema": Array [
                      "bucket",
                    ],
                  },
                  "function": "aggFilter",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "customMetric": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "emptyAsNull": Array [
                      false,
                    ],
                    "enabled": Array [
                      true,
                    ],
                    "field": Array [
                      "bytes",
                    ],
                    "id": Array [
                      "4-metric",
                    ],
                    "schema": Array [
                      "metric",
                    ],
                  },
                  "function": "aggCardinality",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "enabled": Array [
            true,
          ],
          "id": Array [
            "4",
          ],
          "schema": Array [
            "metric",
          ],
        },
        "function": "aggFilteredMetric",
        "type": "function",
      },
    ],
    "type": "expression",
  },
  Object {
    "chain": Array [
      Object {
        "arguments": Object {
          "customBucket": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "enabled": Array [
                      true,
                    ],
                    "filter": Array [
                      Object {
                        "chain": Array [
                          Object {
                            "arguments": Object {
                              "q": Array [
                                "\\"geo.dest: \\\\\\"AL\\\\\\" \\"",
                              ],
                            },
                            "function": "lucene",
                            "type": "function",
                          },
                        ],
                        "type": "expression",
                      },
                    ],
                    "id": Array [
                      "6-filter",
                    ],
                    "schema": Array [
                      "bucket",
                    ],
                  },
                  "function": "aggFilter",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "customMetric": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "emptyAsNull": Array [
                      false,
                    ],
                    "enabled": Array [
                      true,
                    ],
                    "field": Array [
                      "bytes",
                    ],
                    "id": Array [
                      "6-metric",
                    ],
                    "schema": Array [
                      "metric",
                    ],
                  },
                  "function": "aggCardinality",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "enabled": Array [
            true,
          ],
          "id": Array [
            "6",
          ],
          "schema": Array [
            "metric",
          ],
        },
        "function": "aggFilteredMetric",
        "type": "function",
      },
    ],
    "type": "expression",
  },
  Object {
    "chain": Array [
      Object {
        "arguments": Object {
          "customBucket": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "enabled": Array [
                      true,
                    ],
                    "filter": Array [
                      Object {
                        "chain": Array [
                          Object {
                            "arguments": Object {
                              "q": Array [
                                "geo.dest: \\"AL\\" ",
                              ],
                            },
                            "function": "kql",
                            "type": "function",
                          },
                        ],
                        "type": "expression",
                      },
                    ],
                    "id": Array [
                      "8-filter",
                    ],
                    "schema": Array [
                      "bucket",
                    ],
                    "timeWindow": Array [
                      "1m",
                    ],
                  },
                  "function": "aggFilter",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "customMetric": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "emptyAsNull": Array [
                      false,
                    ],
                    "enabled": Array [
                      true,
                    ],
                    "field": Array [
                      "bytes",
                    ],
                    "id": Array [
                      "8-metric",
                    ],
                    "schema": Array [
                      "metric",
                    ],
                  },
                  "function": "aggCardinality",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "enabled": Array [
            true,
          ],
          "id": Array [
            "8",
          ],
          "schema": Array [
            "metric",
          ],
        },
        "function": "aggFilteredMetric",
        "type": "function",
      },
    ],
    "type": "expression",
  },
  Object {
    "chain": Array [
      Object {
        "arguments": Object {
          "customBucket": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "enabled": Array [
                      true,
                    ],
                    "filter": Array [
                      Object {
                        "chain": Array [
                          Object {
                            "arguments": Object {
                              "q": Array [
                                "bytes: *",
                              ],
                            },
                            "function": "kql",
                            "type": "function",
                          },
                        ],
                        "type": "expression",
                      },
                    ],
                    "id": Array [
                      "10-filter",
                    ],
                    "schema": Array [
                      "bucket",
                    ],
                  },
                  "function": "aggFilter",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "customMetric": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "enabled": Array [
                      true,
                    ],
                    "field": Array [
                      "bytes",
                    ],
                    "id": Array [
                      "10-metric",
                    ],
                    "schema": Array [
                      "metric",
                    ],
                    "size": Array [
                      1,
                    ],
                    "sortField": Array [
                      "timestamp",
                    ],
                    "sortOrder": Array [
                      "desc",
                    ],
                  },
                  "function": "aggTopMetrics",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "enabled": Array [
            true,
          ],
          "id": Array [
            "10",
          ],
          "schema": Array [
            "metric",
          ],
        },
        "function": "aggFilteredMetric",
        "type": "function",
      },
    ],
    "type": "expression",
  },
  Object {
    "chain": Array [
      Object {
        "arguments": Object {
          "customBucket": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "enabled": Array [
                      true,
                    ],
                    "filter": Array [
                      Object {
                        "chain": Array [
                          Object {
                            "arguments": Object {
                              "q": Array [
                                "bytes: *",
                              ],
                            },
                            "function": "kql",
                            "type": "function",
                          },
                        ],
                        "type": "expression",
                      },
                    ],
                    "id": Array [
                      "11-filter",
                    ],
                    "schema": Array [
                      "bucket",
                    ],
                  },
                  "function": "aggFilter",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "customMetric": Array [
            Object {
              "chain": Array [
                Object {
                  "arguments": Object {
                    "enabled": Array [
                      true,
                    ],
                    "field": Array [
                      "bytes",
                    ],
                    "id": Array [
                      "11-metric",
                    ],
                    "schema": Array [
                      "metric",
                    ],
                    "size": Array [
                      1,
                    ],
                    "sortField": Array [
                      "timestamp",
                    ],
                    "sortOrder": Array [
                      "desc",
                    ],
                  },
                  "function": "aggTopMetrics",
                  "type": "function",
                },
              ],
              "type": "expression",
            },
          ],
          "enabled": Array [
            true,
          ],
          "id": Array [
            "11",
          ],
          "schema": Array [
            "metric",
          ],
        },
        "function": "aggFilteredMetric",
        "type": "function",
      },
    ],
    "type": "expression",
  },
]
`;
